### ISSUE TITLE
Make the metrics registry configurable

### ISSUE THREAD

**Issue by emschwartz** [autometrics-dev/autometrics-rs#20]

**Comment by emschwartz** [comment]
A number of people have asked me about this so I wanted to follow up with more details on it.

First, to clarify how this works currently with each of the underlying metrics libraries:
- For `opentelemetry`, we call [`opentelemetry_api::global::meter("")`](https://docs.rs/opentelemetry_api/latest/opentelemetry_api/global/fn.meter.html). Applications need to call [`set_meter_provider`](https://docs.rs/opentelemetry_api/latest/opentelemetry_api/global/fn.set_meter_provider.html) for this to have an effect. We call another method that calls this one in the `autometrics::global_metrics_exporter` function.
    - Also note that `opentelemetry` uses the `prometheus` crate under the hood, so we set up the opentelemetry exporter to use the default registry provided by the `prometheus` crate
- For `prometheus`, we encode metrics to the provided [default registry](https://docs.rs/prometheus/latest/prometheus/fn.default_registry.html). This is included by default and doesn't need any special initialization.
- For `metrics`, they have a [`Recorder`](https://docs.rs/metrics/latest/metrics/trait.Recorder.html) trait and a [`set_boxed_recorder`](https://docs.rs/metrics/latest/metrics/fn.set_boxed_recorder.html) function to install that globally. We use the [`metrics_exporter_prometheus::PrometheusBuilder::install`](https://docs.rs/metrics-exporter-prometheus/latest/metrics_exporter_prometheus/struct.PrometheusBuilder.html#method.install) method to set this globally in the `autometrics::global_metrics_exporter` initialization function.

I had initially imagined allowing the registry to be configured for each function (`#[autometrics(registry = REGISTRY)]`). However, someone recently pointed out that this would not only be a bit tedious but actually error-prone because you might forget to pass that argument to some invocations.

A better alternative is to allow the registry to be configured globally for all calls to `autometrics`. However, what we actually need this to do will be slightly different for each metrics library we're using.

- `opentelemetry` - you can already set your own registry if you don't use our `prometheus_exporter` helper functions
- `metrics` - you can also already set your own registry if you don't use our `prometheus_exporter` helper functions
- `prometheus` - we would need to change all of our calls to the macros like [`register_int_counter_vec!`](https://docs.rs/prometheus/latest/prometheus/macro.register_int_counter_vec.html) to the `_with_registry!` variants. We would have our own `static` that is the registry and as long as the method to set the global registry is called before the `Lazy` metrics are initialized, it would use the correct registry. This crate exposes a [`Registry`](https://docs.rs/prometheus/latest/prometheus/struct.Registry.html) struct
- [`prometheus_client`](#25) - this crate exposes a [`Registry`](https://docs.rs/prometheus-client/latest/prometheus_client/registry/struct.Registry.html) struct. Similar to the `prometheus` crate, we would have a global static for the registry and it would work as long as the `set_global_registry` function is called before our metrics are lazily initialized

Something I'm not sure about is what the type signature for a `set_global_registry` function would be. `autometrics` determines which metrics library to use based on the feature flags you have set. It seems a bit odd for the signature to depend on feature flags. 

We could create our own `Registry` trait that's implemented for [`T: metrics::Recorder`](https://docs.rs/metrics/latest/metrics/trait.Recorder.html), [`prometheus::Registry`](https://docs.rs/prometheus/latest/prometheus/struct.Registry.html), and [`prometheus_client::registry::Registry`](https://docs.rs/prometheus-client/latest/prometheus_client/registry/struct.Registry.html). But it also seems a bit odd that the version it would actually expect would be based on the feature flags. We'd need to ensure that you don't accidentally set the wrong type if you have multiple of the metric library feature flags enabled.

... 🤔 

In short, I think this is a reasonable feature request but I'm a bit puzzled about how to implement it.

**Comment by emschwartz** [comment]
If we allow you to pass in the registry, maybe we can also get rid of the [`backends` module](https://github.com/autometrics-dev/autometrics-rs/blob/8a6c50d633bc31e484b5ebab529dbbfb54a9f2c8/autometrics/src/backends/mod.rs) where we expose the `prometheus-client`'s registry. Instead of exposing it, we could just tell users to pass it in if they want to include custom metrics alongside the autometrics metrics.
